refactor: Convert to use `uv`

OCopping opened the pull request:
This PR aims to move away from using the `venv`/`tox` setup for a more modern `uv` (and `tox-uv` until `uv` adds it's own form of runners) approach.

OCopping committed:
Converted to use uv and hatch over venv and tox

GDYendell commented:
If hatch will only be temporary would it be better to stick with tox for now using [tox-uv](https://github.com/tox-dev/tox-uv)?

callumforrester commented:
Relates to #246, just to link issues

OCopping committed:
Replace tox with tox-uv; Remove hatch.yml

OCopping committed:
Fix CI when building env

OCopping committed:
Remove hatch run from _test.yml; Bump pre-commit hook versions

OCopping committed:
Fix pytest runner

OCopping committed:
Update template files for uv

OCopping committed:
Tidy pyproject.toml

GDYendell commented on .pre-commit-config.yaml:
```diff
@@ -1,21 +1,31 @@
         name: lint with ruff
         language: system
-        entry: ruff check --force-exclude --fix
+        entry: ruff check --force-exclude
```
Is this intentional? I find it useful that pre-commit makes the fixes for you.

callumforrester replied:
As a matter of fact I really don't like it/was planning to raise an issue about it, but I don't think the change should go in this PR either way.

OCopping replied:
@callumforrester you want to remove pre-commit fixing?

callumforrester replied:
Not in this PR, no. 

I'm not a big fan of it because I forget it's there, all of our other automated checks don't have side effects so I keep forgetting this one does . So I commit, get an error message from the pre-commit, run `ruff check --fix` and it says "no problems found". 

What are other people's workflows? Commit, see error, notice the word "fixed", commit again, all happy?

GDYendell replied:
I see what you mean, that could be confusing. For me, in lazygit I run commit, get the error and see new unstaged changes appear on the files, so it is really obvious what has happened. But, that won't be the case for everyone. If the prompt at least had an unstaged lines counter that might help.

callumforrester commented on Dockerfile:
```diff
@@ -5,9 +5,18 @@ FROM python:${PYTHON_VERSION} AS developer
 # Add any system dependencies for the developer/build environment here
 RUN apt-get update && apt-get install -y --no-install-recommends \
-    graphviz \
+    graphviz vim \
```
Should: Similar to the above discussion: Potentially useful change, but should go in a different PR for cleanliness reasons.

OCopping replied:
I added this as pre-commit was complaining it couldn't find an editor. Both `nano` and `vi` are not in the image for some reason. So I couldn't actually commit anything.

callumforrester replied:
Interesting... it just uses vscode as the editor for me?

callumforrester commented on .devcontainer/devcontainer.json:
```diff
@@ -41,6 +41,6 @@
-    // After the container is created, install the python project in editable form
-    "postCreateCommand": "pip install $([ -f dev-requirements.txt ] && echo '-c dev-requirements.txt') -e '.[dev]' && pre-commit install"
+    // After the container is created, install the uv env
+    "postCreateCommand": "uv pip install $([ -f dev-requirements.txt ] && echo '-c dev-requirements.txt') -e '.[dev]' && pre-commit install"
```
Must: Not sure if this has already been discussed but needing to prefix every `pip` command with `uv` is a pretty large breaking change. Do we want to commit to it and make a major release or is there some aliasing we could do? @GDYendell and @coretl

OCopping replied:
yeah `uv pip` works exactly the same but is waaaay faster. Worst case you just alias it?

callumforrester commented on .github/workflows/ci.yml:
```diff
@@ -18,30 +18,38 @@ jobs:
     secrets:
       CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
 
-  docs:
```
Is this intentional?

OCopping replied:
Ah this was a mistake, must have overwritten the whole file instead of just copying a few minor changes

OCopping committed:
Revert unintended ci.yml changes

OCopping committed:
Fix actions build, and updates to pyproject.toml

OCopping committed:
Improvements to template pyproject.toml

callumforrester commented:
@OCopping Should probably also update the tests to use `uv pip` now too: https://github.com/DiamondLightSource/python-copier-template/blob/002a1c5cccd235b4898c02b74cec63f6265d7adc/tests/test_example.py

OCopping commented:
> @OCopping Should probably also update the tests to use `uv pip` now too: https://github.com/DiamondLightSource/python-copier-template/blob/002a1c5cccd235b4898c02b74cec63f6265d7adc/tests/test_example.py

I have some of those changes uncommited locally; was trying to fix the CI first

OCopping committed:
Fixes to dev container and tests (but some test cases fail, e.g. running test_template_mypy

coretl reviewed (commented):
I got half way through a review a while back, not sure if these are still relevant, but I'll submit them now anyway. I'll try and find time to do a proper review next week

callumforrester commented on Dockerfile:
```diff
@@ -5,9 +5,18 @@ FROM python:${PYTHON_VERSION} AS developer
+    graphviz vim \
     && rm -rf /var/lib/apt/lists/*
+# Install uv using the official installer script
+RUN curl -LsSf https://astral.sh/uv/install.sh | \
+    env UV_INSTALL_DIR="/usr/local/bin" sh
```
I would say changing the base image for every downstream project is a bigger change and more likely to result in unintended consequences for the sake of removing 1 line from the Dockerfile

callumforrester commented on .pre-commit-config.yaml:
```diff
@@ -1,21 +1,31 @@
       - id: end-of-file-fixer
 
-  - repo: local
+  - repo: https://github.com/astral-sh/uv-pre-commit
+    # uv version.
+    rev: 0.6.14
     hooks:
+      - id: uv-lock
```
What would your preference be? Update only on tags? Commits to `main`?

OCopping replied:
Would you prefer for a `sync` instead?
```yaml
      # Synchronize dependencies upon branch checkout, pull or rebase
      # requires extra hook types to be installed!!!
      - id: uv-sync
```

GDYendell replied:
Doesn't it only update the lockfile if the dependencies in the local environment have changed? Or does lock pull in newer versions itself? That seems like a problem because committing could break things by updating dependencies.

coretl replied:
My preference is to separate the lockfile updating out into a separate periodic job. So pre-commit doesn't do anything with the lockfile, it requires the developer to do something (`uv add my-new-dep`), or wait for the weekly dependabot/renovate nag that updates everything to the latest and runs the tests

callumforrester replied:
Strongly in favour of renovate
https://docs.astral.sh/uv/guides/integration/dependency-bots/

OCopping commented on .github/actions/install_requirements/action.yml:
```diff
@@ -20,15 +17,25 @@ runs:
+      run: uv pip list
+      shell: bash
+
+    - name: Add venv path to Github environment
+      run: echo "VENV_PATH=.venv" >> $GITHUB_ENV
+      shell: bash
+
+    - name: Add venv path to Github Path
+      run: echo "${{ env.VENV_PATH }}/bin:$PATH" >> $GITHUB_PATH
```
Without this GH Actions steps run after this wouldn't know where the venv is located, so would fail.

OCopping committed:
Revert changes to ci.yml

OCopping committed:
Update pre-commit-config

callumforrester committed:
Revert discussed changes and remove commented out parts

callumforrester commented:
From discussion with @coretl we'd like to make an alpha release/a few alpha releases from this branch (`4.0.0a1,2,3...`) and trial them in 1-2 projects before wider release.

callumforrester committed:
Move to managing devcontainer environment outside of dockerfile

coretl commented on .devcontainer/devcontainer.json:
Can delete the `defaultInterpreterPath` now as vscode knows to look in `.venv`

coretl commented on .github/actions/install_requirements/action.yml:
```diff
@@ -28,17 +28,8 @@ runs:
-    - name: Add venv path to Github Path
-      run: echo "${{ env.VENV_PATH }}/bin:$PATH" >> $GITHUB_PATH
-      shell: bash
+      run: uv pip freeze
```
How about [`uv tree`](https://docs.astral.sh/uv/reference/cli/#uv-tree)?

callumforrester committed:
Add config for using uv

Add config for using uv to pyproject.toml and
templated pyproject.toml, add uv lockfile.

…

Co-authored-by: Oliver Copping <oliver.copping@diamond.ac.uk>

callumforrester committed:
Update Dockerfile and devcontainer to use uv

Update Dockerfile based on uv docs, include
uv build in devcontainer, moving from settings
from Dockerfile into devcontainer.json.

…

Co-authored-by: Oliver Copping <oliver.copping@diamond.ac.uk>

callumforrester committed:
Update workflows to use uv

Update Github workflows to use uv

…

Co-authored-by: Oliver Copping <oliver.copping@diamond.ac.uk>

OCopping commented:
> From discussion with @coretl we'd like to make an alpha release/a few alpha releases from this branch (`4.0.0a1,2,3...`) and trial them in 1-2 projects before wider release.
> 
> Tagging @OCopping and @AlexanderWells-diamond for awareness

Are we going to do an alpha release? I would love to try this in one of my projects without having to manually copy the changes over :sweat_smile:

callumforrester commented:
@OCopping we can if you like, but equally you can just do `copier copy --ref move-to-uv ...` and it will just take the commits from this branch, so I'm not sure if it's worth doing a release

coretl commented on .devcontainer/devcontainer.json:
```diff
@@ -39,8 +40,14 @@
-    // After the container is created, install the python project in editable form
-    "postCreateCommand": "pip install $([ -f dev-requirements.txt ] && echo '-c dev-requirements.txt') -e '.[dev]' && pre-commit install"
+    // After the container is created, install the uv env
+    "postCreateCommand": "uv sync && pre-commit install"
```
Docs state that dev is synced by default:
https://docs.astral.sh/uv/concepts/projects/sync/#syncing-development-dependencies

OCopping replied:
Oh weird, dunno what happened on my dev container build then...

coretl commented on pyproject.toml:
```diff
@@ -32,27 +33,51 @@ addopts = """
+env_list = ["pre-commit", "tests", "docs"]
+
+[tool.tox.env]
+passenv = { "*" = "" }
```
I'm trying this out in ophyd-async and I think we should stick with the existing `tox-direct` rather than moving to `tox-uv` which seems to come with its own ideas of which uv to use. Hopefully we will move to a uv task runner soon, if not we can revisit the task runner conversation in a separate PR

OCopping committed:
Fix Dockerfiles (move UV_VERSION to top of file, and assign new build step)

OCopping commented on template/pyproject.toml.jinja:
```diff
@@ -77,30 +79,78 @@ data_file = "/tmp/{{ package_name }}.coverage"
+skipsdist = true
+requires = ["tox-uv>=1.25.0"]
+# envs to runs automatically with tox -p
+env_list = ["pre-commit", "type-checking", "docs"] # "tests"]
```
Oh maybe, but the comment was also to highlight that it's commented out in the .jinja file

OCopping commented on .devcontainer/devcontainer.json:
```diff
@@ -39,8 +40,14 @@
-    // After the container is created, install the python project in editable form
-    "postCreateCommand": "pip install $([ -f dev-requirements.txt ] && echo '-c dev-requirements.txt') -e '.[dev]' && pre-commit install"
+    // After the container is created, install the uv env
+    "postCreateCommand": "uv sync && pre-commit install"
```
Ah wait, we are using a "dependency-group" which does sync dev by default, but using "optional-dependency" does not do this. My bad...

OCopping committed:
Convert _dist.yml to use uvx over pipx

OCopping committed:
Update to use setup-uv@v6

OCopping committed:
Convert to use dls ubuntu-devcontainer

gilesknap committed:
fix apt installs in Dockerfile template

OCopping committed:
Add missing if docker jinja tag

coretl committed:
Merge remote-tracking branch 'origin/main' into move-to-uv

coretl committed:
Add in changes from ophyd-async

coretl committed:
Merge remote-tracking branch 'origin/move-to-uv' into move-to-uv

coretl committed:
tox-uv tweaks

coretl committed:
Disable pep8-naming for the top level

Otherwise ruff complains, the template tests catch this anyway

coretl commented:
Still need to update the docs

coretl committed:
Add docs

AlexanderWells-diamond reviewed (dismissed):
Overall this looks fine to me; most of my comments are just minor questions/clarifications. 
I tried coverting `aioca` to use this version of the template and it appears to have run as expected, with failures explainable as issues with that project, not the template itself: https://github.com/DiamondLightSource/aioca/pull/74

coretl committed:
Update .github/workflows/_tox.yml

Co-authored-by: Alexander Wells <79699091+AlexanderWells-diamond@users.noreply.github.com>

coretl committed:
Address review comments

coretl commented on tests/test_example.py:
```diff
@@ -37,10 +37,16 @@ def run_pipe(cmd: str, cwd=None):
 
 
 def make_venv(project_path: Path) -> callable:
```
Fixed (hopefully)

coretl commented on template/Dockerfile.jinja:
```diff
@@ -1,23 +1,23 @@
-# Set up a virtual environment and put it in PATH
-RUN python -m venv /venv
-ENV PATH=/venv/bin:$PATH{% if docker %}
+    graphviz{% if docker %}
```
Yes, added

coretl commented on Dockerfile:
```diff
@@ -1,13 +1,7 @@
-# Set up a virtual environment and put it in PATH
-RUN python -m venv /venv
-ENV PATH=/venv/bin:$PATH
+    graphviz
```
Yes, it saves 32MB which is more than I thought...

AlexanderWells-diamond commented on template/{% if git_platform=="github.com" %}.github{% endif %}/workflows/ci.yml.jinja:
```diff
@@ -20,10 +20,6 @@ jobs:
       matrix:
         runs-on: ["ubuntu-latest"] # can add windows-latest, macos-latest
         python-version: ["3.11", "3.12", "3.13"]
-        include:
-          # Include one that runs in the dev environment
-          - runs-on: "ubuntu-latest"
-            python-version: "dev"
```
Right, thanks for the explanation

coretl committed:
Move back to tox TOML format

But add extra docs-autobuild target

coretl committed:
Don't autoforward ports in devcontainer

Otherwise docs autobuild doesn't work as the forward kicks in before sphinx can start the server

coretl committed:
Tidy diff

OCopping commented on template/pyproject.toml.jinja:
```diff
@@ -18,22 +18,21 @@ license.file = "LICENSE"
     {% if sphinx %}"pydata-sphinx-theme>=0.12",
     {% endif %}{% if type_checker=="pyright" %}"pyright",
-    {% endif %}"pytest",
+    {% endif %}"pytest>8.3.4",
```
no clue, so guess it can be removed

OCopping committed:
Remove pytest version pinning from template pyproject.toml

AlexanderWells-diamond reviewed (approved):
Looks good to me
